texture height vertex displacement and mesh tessellation

Xtarsia opened the pull request:
This went through many itterations, but is finally getting close.

Implements:

- Clipmap Mesh subdivision, upto 5 levels, for 32 vertices per meter.
- A Displacement Buffer via Veiwport, that renders 1 pixel per vertex, for each subdivision, upto the full 1m vertices.
- A debug view that renders the buffer in fragment.
- Per asset height offset and scale settings.
- Material option to disable / enable displacement (seperate from the tessellation level mesh setting)
- use propper TNB martix for world space texture normals. Previous version was wrong (needed transforming from view space)

Todo:

- Fix cleanup / crash on quit
- Probably need to check vertex spacing
- fully disable more code when tessellation level is 0
- test mobile
- test web

TokisanGames commented:
You said you want to get #717 in next, before this is finalized?

Saul2022 commented:
Testing with the artifact the  terrain system seems to not work at all throwing these errors ``` Godot Engine v4.4.stable.official (c) 2007-present Juan Linietsky, Ariel Manzur & Godot Contributors.
--- Debug adapter server started on port 6006 ---
--- GDScript language server started on port 6005 ---
  ERROR: godot-cpp/src/godot.cpp:333 - Cannot load a GDExtension built for Godot 4.4.1 using an older version of Godot (4.4.0).
  ERROR: core/extension/gdextension_library_loader.cpp:233 - GDExtension initialization function 'terrain_3d_init' returned an error.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:20 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:21 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:24 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:399 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:44 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:82 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:109 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:185 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:223 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:229 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:233 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:234 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:373 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:374 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:400 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:414 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: modules/gdscript/gdscript.cpp:3022 - Failed to load script "res://addons/terrain_3d/src/editor_plugin.gd" with error "Parse error".
  ERROR: godot-cpp/src/godot.cpp:333 - Cannot load a GDExtension built for Godot 4.4.1 using an older version of Godot (4.4.0).
  ERROR: core/extension/gdextension_library_loader.cpp:233 - GDExtension initialization function 'terrain_3d_init' returned an error.
  ERROR: res://addons/terrain_3d/src/region_gizmo.gd:37 - Parse Error: Identifier "Terrain3DData" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/region_gizmo.gd:37 - Parse Error: Identifier "Terrain3DData" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/region_gizmo.gd:49 - Parse Error: Identifier "Terrain3DData" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:48 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:52 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:53 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:54 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:55 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:151 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:151 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:530 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:83 - Parse Error: Could not resolve external class member "tool_changed".
  ERROR: res://addons/terrain_3d/src/ui.gd:87 - Parse Error: Could not resolve external class member "picking".
  ERROR: res://addons/terrain_3d/src/ui.gd:99 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:99 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:142 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:143 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:165 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:170 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:174 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:174 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:176 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:180 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:188 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:192 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:204 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:215 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:226 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:226 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:226 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:231 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:267 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:301 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:302 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:305 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:307 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:308 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:310 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:343 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:358 - Parse Error: Identifier "Terrain3DMaterial" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:359 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:365 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:372 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:381 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:385 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:387 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:389 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:402 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:404 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:411 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:418 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:421 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:424 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:427 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:429 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:432 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:435 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:438 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:441 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:444 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:447 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:450 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:453 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:461 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:537 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:541 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:551 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:554 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:554 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:556 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:557 - Parse Error: Identifier "Terrain3DRegion" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:558 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:560 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:562 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:570 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/ui.gd:573 - Parse Error: Could not resolve external class member "pick".
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:20 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:21 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:24 - Parse Error: Could not find type "Terrain3DEditor" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:399 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:44 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:82 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:109 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:185 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:223 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:229 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:233 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:234 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:373 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:374 - Parse Error: Identifier "Terrain3DEditor" not declared in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:400 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/src/editor_plugin.gd:414 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://demo/src/DemoScene.gd:4 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: modules/gdscript/gdscript.cpp:3022 - Failed to load script "res://demo/src/DemoScene.gd" with error "Parse error".
  ERROR: res://addons/terrain_3d/extras/particle_example/terrain_3D_particles.gd:13 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: res://addons/terrain_3d/extras/particle_example/terrain_3D_particles.gd:13 - Parse Error: Cannot use simple "@export" annotation because the type of the initialized value can't be inferred.
  ERROR: res://addons/terrain_3d/extras/particle_example/terrain_3D_particles.gd:125 - Parse Error: Could not find type "Terrain3D" in the current scope.
  ERROR: modules/gdscript/gdscript.cpp:3022 - Failed to load script "res://addons/terrain_3d/extras/particle_example/terrain_3D_particles.gd" with error "Parse error".
  ERROR: Cannot get class 'Terrain3DMeshAsset'.
  ERROR: Cannot get class 'Terrain3DMeshAsset'.
  ERROR: Cannot get class 'Terrain3DTextureAsset'.
  ERROR: Cannot get class 'Terrain3DTextureAsset'.
  ERROR: Cannot get class 'Terrain3DAssets'.
  ERROR: Cannot get class 'Terrain3DMaterial'.
  ERROR: Cannot get class 'Terrain3D'.
```

TokisanGames commented:
Saul, that's your issue. Read the first error and use the correct Godot version, 4.4.1.

Saul2022 commented:
> Saul, that's your issue. Read the first error and use the correct Godot version, 4.4.1.

Yea you right sorry  also the performance here is so funny lol ( this is with tessellation level 5 and its faster than dual scaling 🤣 great job, this could make it to the lightmapper  shader considering how well it performs.

![Screenshot_20250705_115126](https://github.com/user-attachments/assets/d86b13af-60ea-49d3-bf3c-98ab219d3f36)
![Screenshot_20250705_115119](https://github.com/user-attachments/assets/55a9e36d-80a6-42fa-b764-4957a4dcca26)

Xtarsia commented on src/terrain_3d.cpp:
```diff
@@ -260,6 +267,51 @@ void Terrain3D::_destroy_mouse_picking() {
+	add_child(_d_buffer_vp, true);
+	_d_buffer_vp->set_size(Vector2i(2, 2));
+	_d_buffer_vp->set_disable_3d(true);
+	_d_buffer_vp->set_update_mode(SubViewport::UPDATE_ONCE);
```
the buffer viewport is 2D only, with 3D set to disabled. So most of the other settings for the mouse picking have no effect, and are skipped entirley.

I have added the nearest filtering.

TokisanGames commented:
Please let me know whenever you're ready for another look. You can also rebase as I see there are some conflicts, as well as squashing, including fixup my commits.

TokisanGames commented on src/terrain_3d_material.cpp:
```diff
@@ -412,6 +444,23 @@ void Terrain3DMaterial::_update_shader() {
+			_buffer_shader_override->connect("changed", callable_mp(this, &Terrain3DMaterial::_update_shader));
+		}
+	} else {
+		code = _generate_buffer_shader_code();
```
We don't need to generate this if tessellation == 0. We should still set the buffer material with an empty so we can toggle it on and off and have it clear. Check my proposed commit.

TokisanGames commented:
Getting closer. I reviewed the usage and uploaded some commits that you can review. I will look closer at the code next.

Here are some further issues I've found:

* Like the autoshader and dual scale, I think we should disable tessellation by default, but enable it in the demo. I've done so.

* The lightweight shader is now broken when tessellation > 0, like the minimum shader was.

* Displacement doesn't work reliably outside of regions. Tessellation and the displacement buffer do. Go to a mountain top in the world noise and play with displacement offset/scale and it won't do anything. Go to the ground shortly outside a region and swap textures and displacement does work. Go to a mountain top within a region, even a far one, and it works fine.

## Texture Asset

* If texture asset displacement scale is 1 and offset of the cliff texture is changed 0 - 0.85 it moves evenly, but .85 to 1.0 the peaks squish like its clamped. Do we really need that clamp? What if the texture offset parameter allowed great than 1 if they input it?

* If offset is 0, and I slide displacement scale of cliff down to 0.2 it moves evenly, but from .2 to 0 it increases in size. Is there something that should be clamped there because it's going negative? If offset is 1, this doesn't occur.

## Material

* auto_slope, auto height reduction, base/overlay textures now appear twice in the auto shader group. They have the same values, so you need to de-duplicate what you add to the material parameters.
* Vertical projection/projection threshold appeared in the general and auto shader group. Now it appears in general/world noise.
* Displacement sharpness was in a displacement buffer group far down the list. Displacement scale appeared in the autoshader group. I moved both to a displacement group and sorted it.
* With the minimum shader, if tessellation > 0, it shows the vertical projection, auto shader and displacement blocks.

TokisanGames commented on src/terrain_3d.cpp:
```diff
@@ -634,9 +701,10 @@ void Terrain3D::set_vertex_spacing(const real_t p_spacing) {
 			update_region_labels();
 			_instancer->_update_vertex_spacing(_vertex_spacing);
 			_mesher->reset_target_position();
-			_material->_update_maps();
```
This function is and was private. How did we call it? Is there a left over `friend` or `public:`? I don't see the removal of that facility.

TokisanGames commented on src/terrain_3d_material.cpp:
```diff
@@ -479,7 +536,7 @@ void Terrain3DMaterial::_update_shader() {
 }
 
-void Terrain3DMaterial::_update_maps() {
+void Terrain3DMaterial::_update_maps(const RID &p_material) {
```
Terrain3DData::update_maps() actually regenerates the TextureArray maps.
This function could be renamed to _update_uniforms or _update_material perhaps.

TokisanGames commented on src/terrain_3d_material.cpp:
```diff
@@ -596,18 +670,26 @@ void Terrain3DMaterial::destroy() {
-void Terrain3DMaterial::update() {
-	_update_shader();
-	_update_maps();
+void Terrain3DMaterial::update(bool p_full) {
```
Interestingly, this exposed, public function update() even with full, doesn't pass the texture arrays. It regenerates the shader, then passes all of the non-texture array uniforms. For completeness I think it should in case the user does something and needs a full update, though I don't know what.

TokisanGames commented on src/terrain_3d_material.cpp:
```diff
@@ -582,9 +651,14 @@ void Terrain3DMaterial::initialize(Terrain3D *p_terrain) {
 	_update_shader();
-	_update_maps();
+	_update_maps(_material);
+	_update_maps(_buffer_material);
```
This should just call `update(true)`

TokisanGames commented on src/shaders/displacement_buffer.glsl:
```diff
@@ -0,0 +1,376 @@
+
+// Public uniforms
+uniform bool vertical_projection = true;
+uniform float projection_threshold : hint_range(0.0, 0.99, 0.01) = 0.8;
```
You could wrap these in `group_uniforms general;` `group_uniforms;` to at least put them in the right group.

However, that still leaves them displaying twice in the main shader, and they appear on minimum and lightweight.

TokisanGames commented:
Ok I'm finished. Back to you to review my changes and fix up the remaining things.

TokisanGames commented:
Displacement scale could remain in the material as is. Or it could be a private uniform in the shader, set by a setting combined with the other displacement settings. Then everything is in one place.

Xtarsia commented:
> Displacement scale could remain in the material as is. Or it could be a private uniform in the shader, set by a setting combined with the other displacement settings. Then everything is in one place.

Yeah the terrain shader displacement scale is a "global" scale that determines the maximum displacement distance.

Eg if set to 2 (meters) then the maximum extrusion, and depresion is 1m, for a total of 2m.

Making it private and moving to the mesh section as a part  of a subgroup would make sense as well. (as well as adding some blurb about what I just said to the docs)

Xtarsia commented:
Moved the override, and displacement scale into a subgroup under mesh. (not sure how I make it hide if tessellation == 0)

Added the sharpness property update. (it really is much better seeing it in real time)

Dropped the generated type commit.

Added the extra notes to the docs.

TokisanGames commented:
> not sure how I make it hide if tessellation == 0

_get_property_list()

See OOTA/Lod.gd for an example. Add a subgroup there.

Xtarsia commented:
found out about _validate_property() after a bit of searching. Seems to do the trick nicley for this use case. (hiding the displacement group from the inspector)

updated docs a bit more to include your notes about displaying the collision mesh.

re-ordered displacement scale and offset.

moved the displacement scale inspector property from material to terrain3d as well.

Hopefully its ready now.

TokisanGames reviewed (approved):
Ok, looks good. I've adjusted the docs and parameter ordering.

The last remaining issue for this PR is Flat mode texturing is broken with the lightweight shader.
<img width="797" height="632" alt="image" src="https://github.com/user-attachments/assets/11b37882-6d87-4fa4-8f8e-c1f6f911c999" />

Please fix that, then you can merge in this PR.
